<a href="https://colab.research.google.com/github/manusaleal/EDA_Projeto_Gas/blob/desenvolvimento-cicero/EDA_Projeto_Gas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Projeto

Segundo a fonte:

Até 30 de outubro de 2004, todos os preços médios eram calculados por média aritmética simples. **Após esta data, os preços médios de revenda e de distribuição de combustíveis, em nível estadual, regional e nacional, passaram a ser ponderados com base nas informações de vendas enviadas pelas distribuidoras à ANP.** Atualmente, apenas o preço médio em nível municipal é obtido por média aritmética simples.

***Portanto... Analisar apenas dados fornecidos a partir do dia 31/10/2004.***

Tipos de Produtos:
* Gasolina Comum
* Etanol Hidratado Combustível
* Óleo Diesel Não Aditivado
* Gás Natural Veicular (GNV)
* Gás Liquefeito de Petróleo (GLP – botijão de 13 quilos)
* Óleo diesel B S-10 <---> **Foi adicionado depois**

***Os valores NULL foram substituídos pelo valor -99999***

In [568]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [569]:
import pandas as pd
import numpy as np

In [570]:
#Abrindo a base de dados .csv
df_base = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/TPI (PY) - Projeto Final/2004-2021.csv')

#Realizando uma rápida observação dos dados
df_base.head(10)

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
5   2004-05-09  2004-05-15      NORDESTE               BAHIA   
6   2004-05-09  2004-05-15      NORDESTE               CEARA   
7   2004-05-09  2004-05-15      NORDESTE            MARANHAO   
8   2004-05-09  2004-05-15      NORDESTE             PARAIBA   
9   2004-05-09  2004-05-15      NORDESTE          PERNAMBUCO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   
5  ETANOL HIDRATADO                           408              R$/l   
6  ETANOL HIDRATADO                           278              R$/l   
7  ETANOL HIDRATADO                           105              R$/l   
8  ETANOL HIDRATADO                           125              R$/l   
9  ETANOL HIDRATADO                           423              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                 1.190   
1                1.162                  0.114                 0.890   
2                1.389                  0.097                 1.180   
3                1.262                  0.070                 1.090   
4                1.181                  0.078                 1.050   
5                1.383                  0.132                 0.999   
6                1.453                  0.218                 1.030   
7                1.631                  0.158                 1.350   
8                1.284                  0.130                 1.100   
9                1.224                  0.141                 0.989   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   
5                 2.050                0.426                     0.095   
6                 1.950                0.353                     0.150   
7                 2.000                0.515                     0.097   
8                 1.699                0.353                     0.101   
9                 1.700                0.277                     0.115   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   
5                    0.957                      0.128   
6                      1.1                      0.068   
7                    1.116                      0.091   
8                    0.931                      0.072   
9                    0.947                      0.091   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                     01.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4    

##Pré Análise dos Dados

In [571]:
#Verificando o tamanho do nosso dataset
df_base.shape

(120823, 18)

In [572]:
#Verificando as colunas existentes no dataset
df_base.columns

Index(['DATA INICIAL', 'DATA FINAL', 'REGIÃO', 'ESTADO', 'PRODUTO',
       'NÚMERO DE POSTOS PESQUISADOS', 'UNIDADE DE MEDIDA',
       'PREÇO MÉDIO REVENDA', 'DESVIO PADRÃO REVENDA', 'PREÇO MÍNIMO REVENDA',
       'PREÇO MÁXIMO REVENDA', 'MARGEM MÉDIA REVENDA',
       'COEF DE VARIAÇÃO REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO',
       'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO',
       'PREÇO MÁXIMO DISTRIBUIÇÃO', 'COEF DE VARIAÇÃO DISTRIBUIÇÃO'],
      dtype='object')

In [573]:
COLUNAS = ['DATA INICIAL', 'DATA FINAL', 'REGIÃO', 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS',
           'UNIDADE DE MEDIDA', 'PREÇO MÉDIO REVENDA', 'DESVIO PADRÃO REVENDA', 'PREÇO MÍNIMO REVENDA',
           'PREÇO MÁXIMO REVENDA', 'MARGEM MÉDIA REVENDA', 'COEF DE VARIAÇÃO REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO',
           'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO', 'PREÇO MÁXIMO DISTRIBUIÇÃO',
           'COEF DE VARIAÇÃO DISTRIBUIÇÃO']

Não foi possível visualizar em df.head() se o dataset possui células vazias ou nulas. Para isto, iremos utilizar o comando isna(), isnull() e somar por coluna.

In [574]:
#Verificando se o dataset possui células:
#   * NaN se o tipo de dado for numérico
#   * None ou NaN se o tipo de dado for objeto
#   * NaT se o tipo de dado for data
#df_base.isna().sum()

#Verificando se o dataset possui células:
#   * None ou numpy.NaN.

#OBS: Células do tipo '' ou numpy.inf não são mapadas por essa função a menos que o comando abaixo seja rodado:
#pd.options.mode.use_inf_as_na = True
df_base.isnull().sum()

DATA INICIAL                     0
DATA FINAL                       0
REGIÃO                           0
ESTADO                           0
PRODUTO                          0
NÚMERO DE POSTOS PESQUISADOS     0
UNIDADE DE MEDIDA                0
PREÇO MÉDIO REVENDA              0
DESVIO PADRÃO REVENDA            0
PREÇO MÍNIMO REVENDA             0
PREÇO MÁXIMO REVENDA             0
MARGEM MÉDIA REVENDA             0
COEF DE VARIAÇÃO REVENDA         0
PREÇO MÉDIO DISTRIBUIÇÃO         0
DESVIO PADRÃO DISTRIBUIÇÃO       0
PREÇO MÍNIMO DISTRIBUIÇÃO        0
PREÇO MÁXIMO DISTRIBUIÇÃO        0
COEF DE VARIAÇÃO DISTRIBUIÇÃO    0
dtype: int64

In [575]:
#Verificando se o dataset possui células:
#   * None ou numpy.NaN ou '' ou nump.inf.
pd.options.mode.use_inf_as_na = True
df_base.isnull().sum()

DATA INICIAL                     0
DATA FINAL                       0
REGIÃO                           0
ESTADO                           0
PRODUTO                          0
NÚMERO DE POSTOS PESQUISADOS     0
UNIDADE DE MEDIDA                0
PREÇO MÉDIO REVENDA              0
DESVIO PADRÃO REVENDA            0
PREÇO MÍNIMO REVENDA             0
PREÇO MÁXIMO REVENDA             0
MARGEM MÉDIA REVENDA             0
COEF DE VARIAÇÃO REVENDA         0
PREÇO MÉDIO DISTRIBUIÇÃO         0
DESVIO PADRÃO DISTRIBUIÇÃO       0
PREÇO MÍNIMO DISTRIBUIÇÃO        0
PREÇO MÁXIMO DISTRIBUIÇÃO        0
COEF DE VARIAÇÃO DISTRIBUIÇÃO    0
dtype: int64

Todas as colunas não possuem valores nulos.

In [576]:
#Seguindo recomendação da fonte as datas deverão ser a partirdo dia 31/10/2004
filtro_data = df_base['DATA INICIAL'] >= '2004-10-31'
df_base = df_base[filtro_data]

#Verificando o tamanho do nosso dataset após o filtro
df_base.shape

(117734, 18)

Com a aplicação do filtro, foram removidas 3089 linhas.

In [577]:
#Analisando a quantidade de valores únicos para as colunas
#'DATA INICIAL', 'DATA FINAL', 'REGIÃO', 'ESTADO', 'PRODUTO', 'UNIDADE DE MEDIDA'

for i in [0, 1, 2, 3, 4, 6]:
  coluna = df_base[COLUNAS[i]].unique()
  tamanho = len(coluna)
  print(f'{COLUNAS[i]} tem {tamanho} valores únicos')

DATA INICIAL tem 848 valores únicos
DATA FINAL tem 848 valores únicos
REGIÃO tem 5 valores únicos
ESTADO tem 27 valores únicos
PRODUTO tem 9 valores únicos
UNIDADE DE MEDIDA tem 3 valores únicos


In [578]:
df_base['REGIÃO'].value_counts()

NORDESTE        40582
NORTE           27289
SUDESTE         18770
CENTRO OESTE    17024
SUL             14069
Name: REGIÃO, dtype: int64

In [579]:
df_base['ESTADO'].value_counts()

PARAIBA                4694
RIO DE JANEIRO         4694
SAO PAULO              4694
RIO GRANDE DO SUL      4694
BAHIA                  4694
SANTA CATARINA         4693
RIO GRANDE DO NORTE    4693
MINAS GERAIS           4692
ESPIRITO SANTO         4690
PERNAMBUCO             4688
CEARA                  4688
PARANA                 4682
ALAGOAS                4681
SERGIPE                4670
MATO GROSSO DO SUL     4668
MATO GROSSO            4466
AMAZONAS               4314
GOIAS                  4034
PIAUI                  3903
MARANHAO               3871
DISTRITO FEDERAL       3856
PARA                   3855
TOCANTINS              3853
RONDONIA               3850
ACRE                   3848
RORAIMA                3838
AMAPA                  3731
Name: ESTADO, dtype: int64

Os dados estão bem distibuídos pelos estados.

In [580]:
df_base['PRODUTO'].value_counts()

GASOLINA COMUM        22895
GLP                   22886
ETANOL HIDRATADO      22765
ÓLEO DIESEL           20519
GNV                   14080
ÓLEO DIESEL S10        9113
OLEO DIESEL S10        2376
OLEO DIESEL            2351
GASOLINA ADITIVADA      749
Name: PRODUTO, dtype: int64

Precisa dar uma tratada na coluna PRODUTO e como observação, a Gasolina Aditivada não estava mapeada na fonte.

Antes de continuar, vamos agrupar os PRODUTOS que deveriam ser iguais.

In [581]:
df_base['PRODUTO'] = df_base['PRODUTO'].replace(['ÓLEO DIESEL S10'], 'OLEO DIESEL S10')
df_base['PRODUTO'] = df_base['PRODUTO'].replace(['ÓLEO DIESEL'], 'OLEO DIESEL')

df_base['PRODUTO'].value_counts()

GASOLINA COMUM        22895
GLP                   22886
OLEO DIESEL           22870
ETANOL HIDRATADO      22765
GNV                   14080
OLEO DIESEL S10       11489
GASOLINA ADITIVADA      749
Name: PRODUTO, dtype: int64

In [582]:
df_base['UNIDADE DE MEDIDA'].value_counts()

R$/l       80768
R$/13Kg    22886
R$/m3      14080
Name: UNIDADE DE MEDIDA, dtype: int64

Tem que tomar cuidado quando for comparar

##Tratamento dos Dados

In [583]:
#Analisando os tipos de variáveis das colunas
df = df_base.copy()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 117734 entries, 675 to 120822
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   117734 non-null  object 
 1   DATA FINAL                     117734 non-null  object 
 2   REGIÃO                         117734 non-null  object 
 3   ESTADO                         117734 non-null  object 
 4   PRODUTO                        117734 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   117734 non-null  int64  
 6   UNIDADE DE MEDIDA              117734 non-null  object 
 7   PREÇO MÉDIO REVENDA            117734 non-null  float64
 8   DESVIO PADRÃO REVENDA          117734 non-null  float64
 9   PREÇO MÍNIMO REVENDA           117734 non-null  float64
 10  PREÇO MÁXIMO REVENDA           117734 non-null  float64
 11  MARGEM MÉDIA REVENDA           117734 non-null  object 
 12  COEF DE VARIAÇÃO REVENDA    

Note que algumas colunas que deveriam ser numéricas estão como object.

EXEMPLO:
* MARGEM MÉDIA REVENDA
* PREÇO MÉDIO DISTRIBUIÇÃO
* DESVIO PADRÃO DISTRIBUIÇÃO
* PREÇO MÍNIMO DISTRIBUIÇÃO
* PREÇO MÁXIMO DISTRIBUIÇÃO
* COEF DE VARIAÇÃO DISTRIBUIÇÃO

Será feita a conversão das mesmas.

In [584]:
#Convertendo para numérico (float neste caso) e coagindo quando necessário
LISTA_NOME_COLUNAS_CONVERTIDAS = ["MARGEM MÉDIA REVENDA", "PREÇO MÉDIO DISTRIBUIÇÃO", "DESVIO PADRÃO DISTRIBUIÇÃO",
                                  "PREÇO MÍNIMO DISTRIBUIÇÃO", "PREÇO MÁXIMO DISTRIBUIÇÃO", "COEF DE VARIAÇÃO DISTRIBUIÇÃO"]

for i in range(0, len(LISTA_NOME_COLUNAS_CONVERTIDAS)):
  nome = LISTA_NOME_COLUNAS_CONVERTIDAS[i]
  #O coerce foi usado para converter em 'NaN' os casos em que não foi possível fazer o cast.
  #O número a ser convertido não pode ser menor que (np.iinfo(np.int64).min)
  # e nem maior que (np.iinfo(np.uint64).max)
  df[nome] = pd.to_numeric(df[nome],errors='coerce')

In [585]:
pd.options.mode.use_inf_as_na = False
df.isnull().sum()

DATA INICIAL                        0
DATA FINAL                          0
REGIÃO                              0
ESTADO                              0
PRODUTO                             0
NÚMERO DE POSTOS PESQUISADOS        0
UNIDADE DE MEDIDA                   0
PREÇO MÉDIO REVENDA                 0
DESVIO PADRÃO REVENDA               0
PREÇO MÍNIMO REVENDA                0
PREÇO MÁXIMO REVENDA                0
MARGEM MÉDIA REVENDA             3416
COEF DE VARIAÇÃO REVENDA            0
PREÇO MÉDIO DISTRIBUIÇÃO         3385
DESVIO PADRÃO DISTRIBUIÇÃO       3385
PREÇO MÍNIMO DISTRIBUIÇÃO        3385
PREÇO MÁXIMO DISTRIBUIÇÃO        3385
COEF DE VARIAÇÃO DISTRIBUIÇÃO    3385
dtype: int64

In [586]:
pd.options.mode.use_inf_as_na = True
df.isnull().sum()

DATA INICIAL                        0
DATA FINAL                          0
REGIÃO                              0
ESTADO                              0
PRODUTO                             0
NÚMERO DE POSTOS PESQUISADOS        0
UNIDADE DE MEDIDA                   0
PREÇO MÉDIO REVENDA                 0
DESVIO PADRÃO REVENDA               0
PREÇO MÍNIMO REVENDA                0
PREÇO MÁXIMO REVENDA                0
MARGEM MÉDIA REVENDA             3416
COEF DE VARIAÇÃO REVENDA            0
PREÇO MÉDIO DISTRIBUIÇÃO         3385
DESVIO PADRÃO DISTRIBUIÇÃO       3385
PREÇO MÍNIMO DISTRIBUIÇÃO        3385
PREÇO MÁXIMO DISTRIBUIÇÃO        3385
COEF DE VARIAÇÃO DISTRIBUIÇÃO    3385
dtype: int64

In [587]:
df_base['MARGEM MÉDIA REVENDA'].value_counts()

-99999.0    5517
-           3416
0.361        320
0.326        311
0.333        304
            ... 
9.571          1
12.499         1
4.702          1
7.23           1
22.671         1
Name: MARGEM MÉDIA REVENDA, Length: 12782, dtype: int64

São exatamente 3431 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

In [588]:
df_base['PREÇO MÉDIO DISTRIBUIÇÃO'].value_counts()

-99999.0    5508
-           3385
1.283         94
1.684         92
1.681         92
            ... 
36.605         1
32.688         1
42.69          1
32.545         1
51.166         1
Name: PREÇO MÉDIO DISTRIBUIÇÃO, Length: 16942, dtype: int64

São exatamente 3400 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

In [589]:
df_base['DESVIO PADRÃO DISTRIBUIÇÃO'].value_counts()

-99999.0    5508
-           3385
0           3204
0.046       1078
0.056       1040
            ... 
5.446          1
5.848          1
4.874          1
6.056          1
5.442          1
Name: DESVIO PADRÃO DISTRIBUIÇÃO, Length: 6132, dtype: int64

São exatamente 3400 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

In [590]:
df_base['PREÇO MÍNIMO DISTRIBUIÇÃO'].value_counts()

-99999.0    5508
-           3385
24           786
26           454
22.5         442
            ... 
1.1404         1
0.8378         1
1.7685         1
1.9724         1
2.9328         1
Name: PREÇO MÍNIMO DISTRIBUIÇÃO, Length: 22594, dtype: int64

São exatamente 3400 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

In [591]:
df_base['PREÇO MÁXIMO DISTRIBUIÇÃO'].value_counts()

-99999.0    5508
-           3385
32           807
35           722
31           628
            ... 
1.7405         1
1.3779         1
1.5082         1
2.4214         1
3.8237         1
Name: PREÇO MÁXIMO DISTRIBUIÇÃO, Length: 23669, dtype: int64

São exatamente 3400 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

In [592]:
df_base['COEF DE VARIAÇÃO DISTRIBUIÇÃO'].value_counts()

-99999.0    5508
-           3385
0           3261
0.022       3207
0.021       3165
            ... 
0.426          1
0.362          1
0.418          1
0.466          1
0.591          1
Name: COEF DE VARIAÇÃO DISTRIBUIÇÃO, Length: 400, dtype: int64

São exatamente 3400 linhas que nã puderam ser conertidas para numeric. Qual é o significado de '-'? Não há significado =/

Perceba que agora existem varias células vazias. Precisamos tratá-las antes de continuarmos.

In [593]:
df = df.dropna(subset = LISTA_NOME_COLUNAS_CONVERTIDAS)
df.shape

(114318, 18)

Foram removidas 3416 linhas.

In [594]:
pd.options.mode.use_inf_as_na = False
df.isnull().sum()

DATA INICIAL                     0
DATA FINAL                       0
REGIÃO                           0
ESTADO                           0
PRODUTO                          0
NÚMERO DE POSTOS PESQUISADOS     0
UNIDADE DE MEDIDA                0
PREÇO MÉDIO REVENDA              0
DESVIO PADRÃO REVENDA            0
PREÇO MÍNIMO REVENDA             0
PREÇO MÁXIMO REVENDA             0
MARGEM MÉDIA REVENDA             0
COEF DE VARIAÇÃO REVENDA         0
PREÇO MÉDIO DISTRIBUIÇÃO         0
DESVIO PADRÃO DISTRIBUIÇÃO       0
PREÇO MÍNIMO DISTRIBUIÇÃO        0
PREÇO MÁXIMO DISTRIBUIÇÃO        0
COEF DE VARIAÇÃO DISTRIBUIÇÃO    0
dtype: int64

In [595]:
pd.options.mode.use_inf_as_na = True
df.isnull().sum()

DATA INICIAL                     0
DATA FINAL                       0
REGIÃO                           0
ESTADO                           0
PRODUTO                          0
NÚMERO DE POSTOS PESQUISADOS     0
UNIDADE DE MEDIDA                0
PREÇO MÉDIO REVENDA              0
DESVIO PADRÃO REVENDA            0
PREÇO MÍNIMO REVENDA             0
PREÇO MÁXIMO REVENDA             0
MARGEM MÉDIA REVENDA             0
COEF DE VARIAÇÃO REVENDA         0
PREÇO MÉDIO DISTRIBUIÇÃO         0
DESVIO PADRÃO DISTRIBUIÇÃO       0
PREÇO MÍNIMO DISTRIBUIÇÃO        0
PREÇO MÁXIMO DISTRIBUIÇÃO        0
COEF DE VARIAÇÃO DISTRIBUIÇÃO    0
dtype: int64

Lembrando que, segundo a fonte, todos os valores NULL foram trocados por -99999. Iremos agora remover essas linhas.

In [596]:
for i in range(0, len(LISTA_NOME_COLUNAS_CONVERTIDAS)):
  df_remove = df.loc[(df[LISTA_NOME_COLUNAS_CONVERTIDAS[i]] == -99999)]
  df = df.drop(df_remove.index)

df.shape

(108801, 18)

Foram removidas 5517 linhas

Analisando a data, seria interessante trabalhar com o dia, mês e ano em colunas separadas para facilitar a análise.

In [597]:
#Novas colunas com o dia, mês e ano baseadas na coluna "DATA INICIAL" e "DATA FINAL"

   #dia_inicial = pd.DatetimeIndex(df_base['DATA INICIAL']).day
   #dia_final = pd.DatetimeIndex(df_base['DATA FINAL']).day

   #mes_inicial = pd.DatetimeIndex(df_base['DATA INICIAL']).month
   #mes_final = pd.DatetimeIndex(df_base['DATA FINAL']).month

   #ano_inicial = pd.DatetimeIndex(df_base['DATA INICIAL']).year
   #ano_final = pd.DatetimeIndex(df_base['DATA FINAL']).year

LISTA_NOME_COLUNAS_NOVAS = ['DIA INICIAL', 'MES INICIAL', 'ANO INICIAL', 'DIA FINAL', 'MES FINAL', 'ANO FINAL']

lista_colunas = list()

for i in range(0, len(LISTA_NOME_COLUNAS_NOVAS)):
  nome = LISTA_NOME_COLUNAS_NOVAS[i]

  if 'INICIAL' in nome:
    coluna_aux = df['DATA INICIAL']
  elif 'FINAL' in nome:
    coluna_aux = df['DATA FINAL']
  else:
    print('Erro Inesperado')
    continue

  if 'DIA' in nome:
    coluna = pd.DatetimeIndex(coluna_aux).day
  elif 'MES' in nome:
    coluna = pd.DatetimeIndex(coluna_aux).month
  elif 'ANO' in nome:
    coluna = pd.DatetimeIndex(coluna_aux).year
  else:
    print('Erro Inesperado')
    continue

#Adicionando as colunas no DataFrame
#df_base = df_base.assign(dia_inicial = dia_inicial, mes_inicial = mes_inicial, ano_inicial = ano_inicial, dia_final = dia_final, mes_final = mes_final, ano_final = ano_final)

  df.insert(len(df.columns), LISTA_NOME_COLUNAS_NOVAS[i], coluna)

#Visualizando o dataset modificado
df.head(10)

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
675   2004-10-31  2004-11-06  CENTRO OESTE    DISTRITO FEDERAL   
676   2004-10-31  2004-11-06  CENTRO OESTE               GOIAS   
677   2004-10-31  2004-11-06  CENTRO OESTE         MATO GROSSO   
678   2004-10-31  2004-11-06  CENTRO OESTE  MATO GROSSO DO SUL   
679   2004-10-31  2004-11-06      NORDESTE             ALAGOAS   
680   2004-10-31  2004-11-06      NORDESTE               BAHIA   
681   2004-10-31  2004-11-06      NORDESTE               CEARA   
682   2004-10-31  2004-11-06      NORDESTE            MARANHAO   
683   2004-10-31  2004-11-06      NORDESTE             PARAIBA   
684   2004-10-31  2004-11-06      NORDESTE          PERNAMBUCO   

              PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
675  ETANOL HIDRATADO                           146              R$/l   
676  ETANOL HIDRATADO                           378              R$/l   
677  ETANOL HIDRATADO                           195              R$/l   
678  ETANOL HIDRATADO                           172              R$/l   
679  ETANOL HIDRATADO                            98              R$/l   
680  ETANOL HIDRATADO                           422              R$/l   
681  ETANOL HIDRATADO                           262              R$/l   
682  ETANOL HIDRATADO                            96              R$/l   
683  ETANOL HIDRATADO                           127              R$/l   
684  ETANOL HIDRATADO                           408              R$/l   

     PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
675                1.632                  0.088                  1.27  ...   
676                1.412                  0.123                  1.20  ...   
677                1.692                  0.091                  1.45  ...   
678                1.626                  0.066                  1.37  ...   
679                1.479                  0.053                  1.36  ...   
680                1.586                  0.112                  1.39  ...   
681                1.509                  0.123                  1.20  ...   
682                1.646                  0.089                  1.49  ...   
683                1.489                  0.054                  1.39  ...   
684                1.437                  0.124                  1.19  ...   

     DESVIO PADRÃO DISTRIBUIÇÃO  PREÇO MÍNIMO DISTRIBUIÇÃO  \
675                       0.127                    0.90510   
676                       0.091                    0.87301   
677                       0.077                    1.05000   
678                       0.114                    0.94670   
679                       0.077                    1.05020   
680                       0.183                    0.67130   
681                       0.076                    1.06730   
682                       0.103                    1.14320   
683                       0.064                    1.06470   
684                       0.083                    1.05340   

     PREÇO MÁXIMO DISTRIBUIÇÃO  COEF DE VARIAÇÃO DISTRIBUIÇÃO  DIA INICIAL  \
675                     1.6527                          0.102           31   
676                     1.4579                          0.083           31   
677                     1.6271                          0.059           31   
678                     1.4600                          0.093           31   
679                     1.4047                          0.059           31   
680                     1.5481                          0.139           31   
681                     1.5559                          0.055           31   
682                     1.5658                          0.076           31   
683                     1.3777                          0.050           31   
684                     1.6133                          0.066           31   

     MES INICIAL  ANO INICIAL  DIA FINAL  MES FINAL  ANO FINAL  
675    

In [598]:
#Analisando os tipos de variáveis das colunas
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108801 entries, 675 to 115883
Data columns (total 24 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   108801 non-null  object 
 1   DATA FINAL                     108801 non-null  object 
 2   REGIÃO                         108801 non-null  object 
 3   ESTADO                         108801 non-null  object 
 4   PRODUTO                        108801 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   108801 non-null  int64  
 6   UNIDADE DE MEDIDA              108801 non-null  object 
 7   PREÇO MÉDIO REVENDA            108801 non-null  float64
 8   DESVIO PADRÃO REVENDA          108801 non-null  float64
 9   PREÇO MÍNIMO REVENDA           108801 non-null  float64
 10  PREÇO MÁXIMO REVENDA           108801 non-null  float64
 11  MARGEM MÉDIA REVENDA           108801 non-null  float64
 12  COEF DE VARIAÇÃO REVENDA    

In [599]:
#Visualizando algumas estatísticas rápidas do dataset
df.describe()

NÚMERO DE POSTOS PESQUISADOS  PREÇO MÉDIO REVENDA  \
count                 108801.000000        108801.000000   
mean                     231.600381            11.323676   
std                      389.837425            18.788799   
min                        1.000000             0.975000   
25%                       45.000000             2.157000   
50%                      105.000000             2.858000   
75%                      239.000000             3.885000   
max                     4155.000000            99.357000   

       DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
count          108801.000000         108801.000000         108801.000000   
mean                0.686996              9.814044             13.121848   
std                 1.413067             16.017485             22.170968   
min                 0.000000              0.659000              0.999000   
25%                 0.077000              1.989000              2.439000   
50%                 0.121000              2.649000              3.190000   
75%                 0.202000              3.639000              4.399000   
max                10.748000             90.000000            120.000000   

       MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
count         108801.000000             108801.000000   
mean               2.347912                  0.046086   
std                4.637498                  0.024299   
min                0.001000                  0.000000   
25%                0.290000                  0.030000   
50%                0.392000                  0.042000   
75%                0.579000                  0.059000   
max               36.847000                  0.262000   

       PREÇO MÉDIO DISTRIBUIÇÃO  DESVIO PADRÃO DISTRIBUIÇÃO  \
count             108801.000000               108801.000000   
mean                   8.975764                    0.585125   
std                   14.447877                    1.316398   
min                    0.506000                    0.000000   
25%                    1.873000                    0.049000   
50%                    2.458000                    0.080000   
75%                    3.445000                    0.156000   
max                   83.137000                   18.385000   

       PREÇO MÍNIMO DISTRIBUIÇÃO  PREÇO MÁXIMO DISTRIBUIÇÃO  \
count              108801.000000              108801.000000   
mean                    7.795636                  10.268345   
std                    12.451593                  16.866790   
min                     0.363200                   0.599200   
25%                     1.731800                   2.017000   
50%                     2.310100                   2.637300   
75%                     3.283400                   3.653400   
max                    83.000000                  94.500000   

       COEF DE VARIAÇÃO DISTRIBUIÇÃO    DIA INICIAL    MES INICIAL  \
count                  108801.000000  108801.000000  108801.000000   
mean                        0.045875      15.662908       6.472468   
std                         0.041208       8.822861       3.462713   
min                         0.000000       1.000000       1.000000   
25%                         0.021000       8.000000       3.000000   
50%                         0.031000      16.000000       6.000000   
75%                         0.059000      23.000000      10.000000   
max                         0.780000      31.000000      12.000000   

         ANO INICIAL      DIA FINAL      MES FINAL      ANO FINAL  
count  108801.000000  108801.000000  108801.000000  108801.000000  
mean     2012.563377      15.665224       6.479297    2012.579223  
std         4.568194       8.786065       3.463446       4.568059  
min      2004.000000       1.000000       1.000000    2004.000000  
25%      2009.000000       8.000000       3.000000    2009.000000  
50%      2013.000000      16.000000       6.000000    2013.000000  
75%      2017.000000    